File: quick sort/cpp_quick_sort/quick_sort.cpp
#include "quick_sort.hpp"
using namespace std;

QuickSortArray::QuickSortArray(int size):
	total_(0)
{
	array_.resize(size);
}

void QuickSortArray::CheckCorrect() {
	bool correct = true;

	for (int i = 0; i < array_.size()-1; i++) {
		if (array_[i] > array_[i+1]) {
			correct = false;
		}
	}
	cout << (correct ? "Ture" : "False");
}

void QuickSortArray::Insert(int number) {
	array_[total_++] = number;
}

int QuickSortArray::OrderedWithIn3(int &start, int &end) {
	int length = end - start + 1;
	int middle = end;

	switch (length) {
		case 1:
			break;
		case 2:
			if (array_[start] > array_[end]) {
				Swap(start, end);
			}
			break;
		case 3:
			middle = (start + end) / 2;

			if (array_[start] > array_[end]) {
				Swap(start, end);
			}
			if (array_[middle] > array_[end]) {
				Swap(middle, end);
			}
			if (array_[start] > array_[middle]) {
				Swap(start, middle);
			}
			break;
	}
	return middle;
}

int QuickSortArray::Partition(int &start, int &end, int &pivot) {
	int left = start - 1;
	int right = end;

	while (true) {
		while (left < end && array_[++left] < array_[pivot]) {
		}
		while (right > start && array_[--right] > array_[pivot]) {
		}

		if (left >= right) {
			break;
		} else {
			Swap(left, right);
		}
	}
	Swap(left, pivot);
	return left;
}

void QuickSortArray::QuickSort() {
	Sort(0, array_.size()-1);
}

void QuickSortArray::Show() {
	for (int i = 0; i < array_.size(); i++) {
		cout << array_[i] << (i == array_.size()-1 ? "" : " ");
	}
	cout << endl;
}

void QuickSortArray::Sort(int start, int end) {
	int median = OrderedWithIn3(start, end);

	if (end - start >= 0) {
		int partition = Partition(start, end, median);
		Sort(start, partition-1);
		Sort(partition+1, end);
	}
}

void QuickSortArray::Swap(int &left, int &right) {
	int temp = array_[left];
	array_[left] = array_[right];
	array_[right] = temp;
}

File: quick sort/python_quick_sort/quick_sort.py
import random

class QuickSortArray():
	def __init__(self):
		self.array = list()

	def CheckCorrect(self):
		correct = True
		for index in range(len(self.array)-1):
			if self.array[index] > self.array[index+1]:
				correct = False

		return correct

	def CleanArray(self):
		self.array.clear()

	def Insert(self, number):
		self.array.append(number)

	def OrderedWithIn3(self, start, end):
		length = end - start + 1
		middle = end

		if length == 2:
			if self.array[start] > self.array[end]:
				self.array[start], self.array[end] = self.array[end], self.array[start]
		elif length == 3:
			middle = int((start + end) / 2)

			if self.array[start] > self.array[end]:
				self.array[start], self.array[end] = self.array[end], self.array[start]
			if self.array[middle] > self.array[end]:
				self.array[middle], self.array[end] = self.array[end], self.array[middle]
			if self.array[start] > self.array[middle]:
				self.array[start], self.array[middle] = self.array[middle], self.array[start]

		self.array[middle], self.array[end] = self.array[end], self.array[middle]
		return middle

	def Partition(self, start, end, pivot):
		left = start
		right = end

		while True:
			for l in range(left, end+1):
				left = l
				if self.array[l] > self.array[pivot]:
					break

			for r in reversed(range(start, right)):
				right = r
				if self.array[r] < self.array[pivot]:
					break

			if left >= right:
				break
			else:
				self.array[left], self.array[right] = self.array[right], self.array[left]

		self.array[left], self.array[pivot] = self.array[pivot], self.array[left]
		return left

	def RandomNumbers(self, numbers):
		for i in range(numbers):
			self.Insert(random.randrange(199))

	def Show(self):
		for index in range(len(self.array)):
			print(str(self.array[index]) if index == len(self.array)-1 else str(self.array[index]) + " ", end='')

		print('')

	def Sort(self, start, end):
		median = self.OrderedWithIn3(start, end)

		if end - start >= 0:
			partition = self.Partition(start, end, median)
			self.Sort(start, partition-1)
			self.Sort(partition+1, end)
